lib/iris/_lazy_data.py

```diff
@@ -45,10 +46,40 @@ def array_masked_to_nans(array, mask=None):
     Convert a masked array to a normal array with NaNs at masked points.
     This is used for dask integration, as dask does not support masked arrays.
     Note that any fill value will be lost.
+
     """
     if mask is None:
         mask = array.mask
     if array.dtype.kind == 'i':
         array = array.astype(np.dtype('f8'))
     array[mask] = np.nan
     return array
+
+
+def array_nans_to_masked(array, fill_value=None, dtype=None, filled=False):
+    """
+    Convert an array into a masked array, by masking any NaN points.
+
+    Optional data type casting is also performed to the specified dtype.
+
+    The resultant array may also be filled with the fill_value, if requested.
+
+    """
+    if (not isinstance(array, ma.masked_array)) and \
+            array.dtype.kind == 'f':
+        # First, calculate the mask.
+        mask = np.isnan(array)
+        # Now, cast the dtype, if required.
+        if dtype is not None and dtype != array.dtype:
+            array = array.astype(dtype)
+        # Finally, mask or fill the data, as required.
+        if np.any(mask):
+            if filled:
+                if fill_value is None:
+                    emsg = 'Invalid fill value, got {!r}.'
+                    raise ValueError(emsg.format(fill_value))
+                array[mask] = fill_value
+            else:
+                array = ma.masked_array(array, mask=mask,
+                                        fill_value=fill_value)
+    return array
```

Comment:
@pp-mo I've combined my `array_nans_to_filled` and `array_nans_to_masked` and opted to use the `filled` kwarg.

Interested on your take with my raising an exception for the `filled == True` case but `fill_value = None` ... seems the right thing to do.

Also, that this routine will perform `dtype` casting ... again, seems to me like the best place to do it, but whether that capability wrapped up within `array_nans_to_masked` is appropriate for a higher level function that calls it, such as your `as_concrete_data`